[18.0][MIG] partner_vat_unique: Migration to 18.0

lef-adhoc opened the pull request.

Commit:
[ADD] partner_vat_unique, check unique VAT number

cubells committed:
[MIG] partner_vat_unique: Migration to 10.0

oca-travis committed:
[UPD] Update partner_vat_unique.pot

michotm committed:
[MIG] partner_vat_unique: Migration to 11.0

oca-travis committed:
[UPD] Update partner_vat_unique.pot

koenloodts committed:
[MIG] partner_vat_unique: Migration to 12.0

oca-travis committed:
[UPD] Update partner_vat_unique.pot

mvrodriguez committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (2 of 2 strings)

Translation: partner-contact-12.0/partner-contact-12.0-partner_vat_unique
Translate-URL: https://translation.odoo-community.org/projects/partner-contact-12-0/partner-contact-12-0-partner_vat_unique/es/

Commit:
Added translation using Weblate (Italian)

Commit:
Translated using Weblate (Italian)

Currently translated at 100.0% (2 of 2 strings)

Translation: partner-contact-12.0/partner-contact-12.0-partner_vat_unique
Translate-URL: https://translation.odoo-community.org/projects/partner-contact-12-0/partner-contact-12-0-partner_vat_unique/it/

pedrocs-exo committed:
Added translation using Weblate (Portuguese)

pedrocs-exo committed:
Translated using Weblate (Portuguese)

Currently translated at 100.0% (2 of 2 strings)

Translation: partner-contact-12.0/partner-contact-12.0-partner_vat_unique
Translate-URL: https://translation.odoo-community.org/projects/partner-contact-12-0/partner-contact-12-0-partner_vat_unique/pt/

sergio-teruel committed:
[FIX] partner_vat_unique: Do not copy vat field

OCA-git-bot committed:
partner_vat_unique 12.0.1.0.1

Commit:
[IMP] partner_vat_unique: black, isort

Commit:
[MIG] partner_vat_unique: Migration to 13.0

oca-travis committed:
[UPD] Update partner_vat_unique.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: partner-contact-13.0/partner-contact-13.0-partner_vat_unique
Translate-URL: https://translation.odoo-community.org/projects/partner-contact-13-0/partner-contact-13-0-partner_vat_unique/

newtratip committed:
[MIG] partner_vat_unique: Migration to 14.0

oca-travis committed:
[UPD] Update partner_vat_unique.pot

pedrobaeza committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (4 of 4 strings)

Translation: partner-contact-14.0/partner-contact-14.0-partner_vat_unique
Translate-URL: https://translation.odoo-community.org/projects/partner-contact-14-0/partner-contact-14-0-partner_vat_unique/es/

alan196 committed:
[MIG] partner_vat_unique: Migration to 15.0

Commit:
[UPD] Update partner_vat_unique.pot

OCA-git-bot committed:
[UPD] README.rst

AnizR committed:
add parent_id in api.constrains of res_partner model

AnizR committed:
[16][MIG] partner_vat_unique: Migration to 16.0

Commit:
[UPD] Update partner_vat_unique.pot

OCA-git-bot committed:
[UPD] README.rst

marylla committed:
Added translation using Weblate (German)

marylla committed:
Translated using Weblate (German)

Currently translated at 100.0% (4 of 4 strings)

Translation: partner-contact-16.0/partner-contact-16.0-partner_vat_unique
Translate-URL: https://translation.odoo-community.org/projects/partner-contact-16-0/partner-contact-16-0-partner_vat_unique/de/

Commit:
[UPD] Update partner_vat_unique.pot

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: partner-contact-16.0/partner-contact-16.0-partner_vat_unique
Translate-URL: https://translation.odoo-community.org/projects/partner-contact-16-0/partner-contact-16-0-partner_vat_unique/

badbole committed:
Added translation using Weblate (Croatian)

badbole committed:
Translated using Weblate (Croatian)

Currently translated at 100.0% (4 of 4 strings)

Translation: partner-contact-16.0/partner-contact-16.0-partner_vat_unique
Translate-URL: https://translation.odoo-community.org/projects/partner-contact-16-0/partner-contact-16-0-partner_vat_unique/hr/

Commit:
[UPD] Update partner_vat_unique.pot

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: partner-contact-16.0/partner-contact-16.0-partner_vat_unique
Translate-URL: https://translation.odoo-community.org/projects/partner-contact-16-0/partner-contact-16-0-partner_vat_unique/

Gelo-fl committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (4 of 4 strings)

Translation: partner-contact-16.0/partner-contact-16.0-partner_vat_unique
Translate-URL: https://translation.odoo-community.org/projects/partner-contact-16-0/partner-contact-16-0-partner_vat_unique/es/

victor-champonnois committed:
Added translation using Weblate (French)

OCA-git-bot committed:
[UPD] README.rst

pedrocs-exo committed:
Translated using Weblate (Portuguese)

Currently translated at 100.0% (4 of 4 strings)

Translation: partner-contact-16.0/partner-contact-16.0-partner_vat_unique
Translate-URL: https://translation.odoo-community.org/projects/partner-contact-16-0/partner-contact-16-0-partner_vat_unique/pt/

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (4 of 4 strings)

Translation: partner-contact-16.0/partner-contact-16.0-partner_vat_unique
Translate-URL: https://translation.odoo-community.org/projects/partner-contact-16-0/partner-contact-16-0-partner_vat_unique/it/

rov-adhoc committed:
[IMP] partner_vat_unique: pre-commit execution

rov-adhoc committed:
[MIG] partner_vat_unique: Migration to 17.0

Commit:
[UPD] Update partner_vat_unique.pot

OCA-git-bot committed:
[BOT] post-merge updates

Commit:
[IMP] partner_contact_role: pre-commit execution

lef-adhoc committed:
[MIG] partner_vat_unique: Migration to 18.0

rousseldenis commented:
/ocabot migration partner_vat_unique

rousseldenis commented on partner_vat_unique/models/res_partner.py:
```diff
@@ -24,5 +24,6 @@ def _check_vat_unique(self):
             if record.same_vat_partner_id:
                 raise ValidationError(
-                    _("The VAT %s already exists in another partner.") % record.vat
+                    self.env._("The VAT %s already exists in another partner.")
```
@lef-adhoc Please use translation function arguments instead.

`self.env._("The VAT %(vat)s already exists in another partner.", vat=record.vat)`

lef-adhoc replied:
Done!

rousseldenis commented on partner_vat_unique/models/res_partner.py:
```diff
@@ -0,0 +1,29 @@
+        for record in self:
+            if record.parent_id or not record.vat:
+                continue
+            test_condition = config["test_enable"] and not self.env.context.get(
```
This is not really good practice. Instead, a configuration parameter is better to enable the constraint.

lef-adhoc replied:
Thanks for the feedback! Do you have an example of how to properly implement this with a configuration parameter?

rousseldenis replied:
In ResConfigSettings model:

```
partner_vat_unique = fields.Boolean(
    config_parameter="partner_vat_unique.partner_vat_unique',
    help="Check this if you want to constrain VAT number to be unique"
)
``` 

Then, here, before the loop:

```
if  self.env["ir.config_parameter"]
            .sudo()
            .get_param("partner_vat_unique.partner_vat_unique', False):
```

rousseldenis replied:
Then, remove all things about test_enable.

And adapt tests accordingly.

Then, put that in a separate commit.

lef-adhoc replied:
I’ve added a commit implementing the configuration parameter for the unique VAT constraint as you suggested. Could you please take a look and let me know if this approach works for you?

lef-adhoc committed:
[IMP] partner_vat_unique: add configuration parameter for unique VAT constraint

- Defined `partner_vat_unique` as a boolean configuration parameter.
- Updated `res.partner` to validate VAT uniqueness based on this parameter.
- Added a new section in ResConfigSettings for enabling/disabling the constraint.
- Updated tests to include scenarios with the configuration parameter.

rousseldenis commented on partner_vat_unique/models/res_partner.py:
```diff
@@ -17,10 +16,11 @@ def _check_vat_unique(self):
-                "test_vat"
-            )
-            if test_condition:
+            if (
```
Put it outside the loop.  If deactivated, don't enter in it at all.

lef-adhoc replied:
Done!

qgroulard commented on partner_vat_unique/models/res_partner.py:
```diff
@@ -14,14 +13,17 @@ class ResPartner(models.Model):
+        if (
+            not self.env["ir.config_parameter"]
+            .sudo()
+            .get_param("partner_vat_unique.partner_vat_unique", default=False)
```
As it's `False` by default, the configuration must be explained in the readme.

qgroulard reviewed (commented):
LGTM, just a small comment.

ced-adhoc reviewed (approved):
LGTM

juancarlosonate-tecnativa reviewed (approved):
LGTM, functional and code review

TT54210

jmartinacu commented:
Hello @pedrobaeza @rousseldenis, could this PR be merge??

JavierVazquezQubiQ commented:
Do you know if this PR will be merge soon? @pedrobaeza @rousseldenis 

Thanks!!

pedrobaeza commented:
/ocabot merge nobump
